File: mr-plugin/server/mr-core/mr_core/models/deployment.py
from __future__ import annotations
from peewee import *
from playhouse.postgres_ext import JSONField
from .base.read_write import ReadWriteModel
# from mr_core.models import machine_configuration
from datetime import datetime, timezone
import uuid
import pytz
# from mr_core.models import metrics


class Deployment(ReadWriteModel):
    id = UUIDField(primary_key=True)
    name = CharField()
    description = TextField(null=True)  # Optional field
    created_at = DateTimeField(null=False, default=lambda: datetime.now(pytz.utc))  # Use DateTimeField for actual date-time values
    updated_at = DateTimeField(null=False, default=lambda: datetime.now(pytz.utc))   # Optional field, use DateTimeField for actual date-time values
    url = CharField()
    model_id = UUIDField()
    user_username = CharField()
    machine_config = JSONField()
    status = CharField(choices=[('pending', 'Pending'), ('active', 'Active'), ('failed', 'Failed'), ('completed', 'Completed')])
    logs_url = CharField(null=True)  # Optional field
    metrics = JSONField()
    deployment_environment = CharField(choices=[('Kubernetes', 'Kubernetes'), ('VM', 'VM'), ('SHPC', 'SHPC')])
    tags = JSONField(null=True)  # Optional array of tags or keywords
    
    @classmethod
    def create(cls, user=None, **query):
        return super(Deployment, cls).create(user, **query)

File: mr-plugin/model-registry-ui/src/api/data_types/deployment.ts
interface MachineConfiguration {
	cpu: string;           // CPU specification (e.g., "4 vCPUs")
	memory: string;        // Memory specification (e.g., "16 GB")
	gpu?: string;          // Optional GPU specification (e.g., "1 GPU")
	disk: string;          // Disk storage specification (e.g., "100 GB SSD")
}

interface Metrics {
	cpu_usage?: number;       // Optional CPU usage percentage
	memory_usage?: number;    // Optional Memory usage percentage
	disk_usage?: number;      // Optional Disk usage percentage
}

export interface DeploymentInterface {
	id: string;                   // Unique identifier for the deployment
	name: string;                 // Name of the deployment
	description?: string;         // Optional description of the deployment
	created_at: string;           // Timestamp when the deployment was created (ISO date string)
	updated_at?: string;          // Optional timestamp of the last update (ISO date string)
	url: string;                  // URL to access the deployed application or model
	model_id: string;             // Unique identifier of the model being deployed
	user_username: string;        // Username of the user who created the deployment
	machine_config: MachineConfiguration; // Machine configuration for the deployment
	status: 'pending' | 'active' | 'failed' | 'completed'; // Current status of the deployment
	logs_url?: string;            // Optional URL to access deployment logs
	metrics?: Metrics;            // Optional metrics such as CPU, memory, and disk usage
	deployment_environment: 'Kubernetes' | 'VM' | 'SHPC'; // Where the deployment is hosted
	tags?: string[];              // Optional array of tags or keywords associated with the deployment
	[key: string]: unknown; // Index signature to handle dynamic properties
}
